### PR TITLE
(ALPHA) growthAnalyze (choice 2 / less vodoo) - only merge one of these

### PR DESCRIPTION
**Pull request by zeddrak**
This implementation involves a simple (and poorly optimized) 50% partition search approximation methodology to determine the number of threads actually required to grow a server from one amount to another.

The advantages of this approach over option 1 is that it allows for the current mechanic of +$1/grow thread pre-multiplier.
The disadvantage is that it can't seamlessly replace the previous functionality (this is just a fact of the current implementation of grow).

The advantage over option 2 is just less math wizardry. Partition searches are a relatively entry level concept compared to using advanced approximation techniques. The disadvantage is that it's almost certainly slower.

### CODE REVIEW

File: src/Server/ServerHelpers.ts [hunk 1]
```diff
@@ -66,6 +66,92 @@ export function numCycleForGrowth(server: Server, growth: number, p: IPlayer, co
   return cycles;
 }
 
+/**
+ * Replacement function for the above function that accounts for the +$1/thread functionality of grow
+ * with parameters that are the same (for compatibility), but functionality is slightly different.
+ * This function can ONLY be used to calculate the threads needed for a given server in its current state,
+ * and so wouldn't be appropriate to use for formulas.exe or ns.growthAnalyze (as those are meant to
+ * provide theoretical scenaarios, or inverse hack respectively). Players COULD use this function with a
+ * custom server object with the correct moneyAvailable and moneyMax amounts, combined with a multplier
```
Comment (tyrope) [inline src/Server/ServerHelpers.ts:75]: typo: multplier => multiplier

File: src/Server/ServerHelpers.ts [hunk 2]
```diff
@@ -66,6 +66,92 @@ export function numCycleForGrowth(server: Server, growth: number, p: IPlayer, co
...
+ * correctly calculated to bring the server to a new moneyAvailable (ie, pasing in moneyAvailable 300 and x2
```
Comment (tyrope) [inline src/Server/ServerHelpers.ts:76]: typo: pasing => passing

File: src/Server/ServerHelpers.ts [hunk 3]
```diff
@@ -66,6 +66,92 @@ export function numCycleForGrowth(server: Server, growth: number, p: IPlayer, co
...
+ * when you want the number of threads required to grow that particular server from 300 to 600), and this
+ * function would pass back the correct number of threads. But the key thing is that it doesn't just
+ * inverse/undo a hack (since the amount hacked from/to matters, not just the multiplier).
+ * The above is also a rather unnecessarily obtuse way of thinking about it for a formulas.exe type of
+ * application, so another function with different parameters is provided for that case below this one.
+ * Instead this function is meant to hand-off from the old numCycleForGrowth function to the new one
+ * where used internally for pro-rating or the like. Where you have applied a grow and want to determine
+ * how many htreads were needed for THAT SPECIFIC grow case using a multiplier.
```
Comment (tyrope) [inline src/Server/ServerHelpers.ts:84]: typo: htreads => threads

File: src/Server/ServerHelpers.ts [hunk 4]
```diff
@@ -66,6 +66,92 @@ export function numCycleForGrowth(server: Server, growth: number, p: IPlayer, co
...
+ * Idealy, this function, and the original function above will be depreciated to use the methodology
```
Comment (tyrope) [inline src/Server/ServerHelpers.ts:85]: typo: Idealy => Ideally

File: src/Server/ServerHelpers.ts [hunk 5]
```diff
@@ -66,6 +66,92 @@ export function numCycleForGrowth(server: Server, growth: number, p: IPlayer, co
...
+export function numCycleForGrowthTransition(server: Server, growth: number, p: IPlayer, cores = 1): number {
+  return numCycleForGrowthCorrected(server, server.moneyAvailable * growth, server.moneyAvailable, p, cores);
+}
+
+/**
+ * This function calculates the number of threads needed to grow a server from one $amount to a the same or higher $amount
+ * (ie, how many threads to grow this server from $200 to $600 for example). Used primarily for a formulas (or possibly growthAnalyze)
+ * type of application. It lets you "theorycraft" and easily ask what-if type questions. It's also the one that implements the
+ * main thread calculation algorith, and so is the fuinction all helper functions should call.
+ * It protects the inputs (so putting in INFINITY for targetMoney will use moneyMax, putting in a nagitive for start will use 0, etc.)
```
Comment (tyrope) [inline src/Server/ServerHelpers.ts:101]: typo: nagitive => negative

**Review by tyrope — COMMENTED** [review]
Code looks good; caught a couple of typos in your comments though. ;)

**Comment by zeddrak** [comment]
Just realized that numCycleForGrowthByHackAmt can be made to function similar to the current growthAnalyze function with the same parameter list. The main difference being that it would return the number of threads needed to reverse the hack amt as if the server were at moneyMax when hacked. Since this is the "worst case" (as well as the most common case), this would mean it returns too many threads for lower value servers, but this has the side effect of slowly growing the server (as the extra threads mean more growth than loss). This would make that function inappropriate for pro-rating, but would make it a very good choice for replacing the current growthAnalyze without breaking scripts (since parameter list is unchanged).

File: src/Server/ServerHelpers.ts [hunk 6]
```diff
@@ -66,6 +66,109 @@ export function numCycleForGrowth(server: Server, growth: number, p: IPlayer, co
...
+ * @param server - Server being grown
+ * @param hackAmt - the amount hacked (total, not per thread) - as a decimal (like 0.60 for hacking 60% of available money)
+ * @param prehackMoney - how much money the server had before being hacked (like 200000 for hacking a server that had $200000 on it at time of hacking)
+ * @param p - Reference to Player object
+ * @returns Number of "growth cycles" needed to reverse the described hack
+ */
+export function numCycleForGrowthByHackAmt(server: Server, hackAmt: number, prehackMoney: number, p: IPlayer, cores = 1): number{
+	if (prehackMoney > server.moneyMax) prehackMoney = server.moneyMax;
+	const posthacMoney = Math.floor(prehackMoney * Math.min(1, Math.max(0, (1 - hackAmt))));
+	return numCycleForGrowthCorrected(server, posthacMoney, prehackMoney, p, cores);
```
Comment (dwRchyngqxs) [inline src/Server/ServerHelpers.ts:152]: I think that swapping the argument here was not correct: we want to know how to grow from a lower value to a greater value, but hacking reduces the money on a server so posthacMony < prehackMoney.
targetMoney should be prehackMoney as before.

Comment (zeddrak) [reply src/Server/ServerHelpers.ts:152]: You are correct... Don't code tired... ;)

**Comment by zeddrak** [comment]
ok, parameters look better now...

**Comment by phyzical** [comment]
tbh im in favour of the simple soultion.

**Pull request closed without merging**